lib/install/rust.js

```diff
@@ -0,0 +1,557 @@
+  var req = request.get(url);
+
+  // When we receive the response
+  req.on('response', (res) => {
+
+    // Parse out the length of the incoming bundle
+    var contentLength = parseInt(res.headers['content-length'], 10);
+
+    // Create a new progress bar
+    var bar = new Progress('     [:bar] :percent :etas remaining', {
+      clear: true,
+      complete: '=',
+      incomplete: ' ',
+      width: 20,
+      total: contentLength
+    });
+
+    // When we get incoming data, update the progress bar
+    res.on('data', (chunk) => {
+      bar.tick(chunk.length);
+    });
+  });
+
+  return req;
+}
+
+function downloadString(url) {
+  return new Promise((resolve, reject) => {
+    request({
+      url,
+      // We want to force Cloudfront to serve us the latest file.
+      headers: {
+        'Accept-Encoding': 'gzip, deflate',
+      },
+    }, (error, response, body) => {
+      if (!error && response.statusCode === 200) {
+        resolve(body);
+      } else {
+        reject(error || response.statusCode);
+      }
+    });
+  });
+}
+
+function tmpdir() {
+  return new Promise((resolve) => {
+    var dir = fsTemp.template('t2-sdk-%s').mkdirSync();
+    resolve({
+      path: dir,
+      cleanup: () => {
+        try {
+          fs.removeSync(dir);
+        } catch (e) {
+          // Swallow errors in removing temporary folder. If the folder was
+          // successfully or unsuccessfully moved, it may not exist at its
+          // location in the temporary directory, but this isn't fatal.
+        }
+      }
+    });
+  });
+}
+
+module.exports.toolchainPath = () => {
+  return new Promise((resolve, reject) => {
+    var sdkPlatformPath = path.join(SDK_PATHS.sdk, getPlatform());
+    var values = fs.readdirSync(sdkPlatformPath);
+
+    for (var i = 0; i < values.length; i++) {
+      if (values[i].match(/^toolchain\-/)) {
+        return resolve(path.join(sdkPlatformPath, values[i]));
+      }
+    }
+    return reject(new Error('No toolchain found.'));
+  });
+};
+
+// Checks if CHECKSUM file in our SDK equals our expected checksum.
+// This will resolve with checking that the SDK exists and matches the checksum.
+module.exports.checkTools = (checksumVerify) => {
+  var dir = path.join(SDK_PATHS.sdk, getPlatform());
+  return new Promise((resolve) => {
+    var checksum = fs.readFileSync(path.join(dir, 'CHECKSUM'), 'utf-8');
+    resolve({
+      exists: true,
+      checked: checksumVerify === checksum,
+      path: dir,
+    });
+  }).catch(() => ({
+    exists: false,
+    checked: false,
+    path: dir,
+  }));
+};
+
+module.exports.checkRustlib = (rustv, checksumVerify) => {
+  var dir = path.join(SDK_PATHS.rustlib, rustv);
+  return new Promise((resolve) => {
+    var checksum = fs.readFileSync(path.join(dir, 'CHECKSUM'), 'utf-8');
+    resolve({
+      exists: true,
+      checked: checksumVerify === checksum,
+      path: dir,
+    });
+  }).catch(() => ({
+    exists: false,
+    checked: false,
```

Comment:
Actually, this is a nice example: this thing _was_ "checked", it just failed for whatever reason.